### PR TITLE
[Go SDK]: Implement mongodbio.Read with an SDF

### LINKED ISSUE #24822: [Task][Go SDK]: Implement mongodbio.Read with a splittable DoFn
### What needs to happen?

Rewrite `mongodbio.Read` to use a splittable DoFn for improved scalability

### Issue Priority

Priority: 2 (default / most normal work should be filed as P2)

### Issue Components

- [ ] Component: Python SDK
- [ ] Component: Java SDK
- [X] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by johannaojeling**
Implements mongodbio.Read with a splittable DoFn. Fixes #24822.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by johannaojeling** [comment]
R: @lostluck

File: sdks/go/pkg/beam/io/mongodbio/read.go [hunk 1]
```diff
@@ -442,53 +266,53 @@ func (fn *readFn) ProcessElement(
...
-	if len(idFilter) == 0 {
-		return customFilter
+		emit(value)
 	}
 
-	if len(customFilter) == 0 {
-		return idFilter
-	}
+	result := cursorResult{isExhausted: true}
+	rt.TryClaim(result)
```
Comment (johannaojeling) [inline sdks/go/pkg/beam/io/mongodbio/read.go:283]: I added this extra `TryClaim` invokation to be able to mark the tracker as done in the event that the cursor is exhausted before the expected count of documents have been claimed. This may happen if documents within the restriction changed, e.g. were deleted or modified to no longer match the filter, since the total number of documents within the restriction was counted upon restriction creation. I preferred to still have the DoFn succeed in this case.

My first idea involved passing only the ID as the pos argument to TryClaim, and returning `sdf.StopProcessing()` after the cursor had been exhausted, inspired by the [example](https://beam.apache.org/documentation/programming-guide/#user-initiated-checkpoint) in the programming guide. However I noticed user initiated checkpointing is less supported among the runners and also my pipeline was interpreted as a streaming job in Dataflow so I was unsure of whether that was the right approach to use and decided to go with this less elegant way instead.

**Review by johannaojeling — COMMENTED** [review]
I hope I have interpreted the SDF, RTracker etc. interfaces correctly and have tried to find a suitable way to implement them for this use case. If something looks strange or you have ideas for improvement, feedback is always welcome.

**Comment by lostluck** [comment]
Very big PR! I didn't have a chance to look at it earlier in the week, and it's larger than I expected, so it may take me a bit to get through it. I will be getting this first pass done in the next few hours however.

File: sdks/go/pkg/beam/io/mongodbio/read.go (hunk 1 above)
Comment (lostluck) [reply sdks/go/pkg/beam/io/mongodbio/read.go:283]: I think you're right in this case. 

Formally, the promise around TryClaim is "if TryClaim returns successfully, I will process the claimed position", so if the underlying data/job mapping, changes to invalidate the restriction, it's reasonable to simply discard previously unknown input, since it wasn't originally there to start with. 

As for process continuations, while they *technically* can be used for batch WRT the global window, you're right that support for them is largely for handling streaming, which is how Dataflow interprets transforms that can return process continuations.  It's possible it's a bug in the Go SDKs interpretation however, and it's the combination of a  Process Continuation and an Unbounded Restriction that should lead to a streaming interpretation but that would need to be vetted against existing runners. The documentation around that part of the model for SDK developers is spotty at best unfortunately.

-------

WRT actual splitting behavior, I know Dataflow is currently the only runner that does dynamic splitting. I believe both Flink and Spark will do the initial splitting stages, which amounts to each element will get a split of their initial restriction after creating the restriction.

This is typically fine. Dynamic splitting or Liquid Sharding shows most of it's benefits in avoiding stragglers so that all the work ends closer together, providing better utilization of workers, and faster completion of a job.

Comment (lostluck) [reply sdks/go/pkg/beam/io/mongodbio/read.go:283]: OK seeing everything in context, I agree 100% this is the correct move because the tracker is managing itself around the external cursor, and can't do the more elegant approach. Ideally, the "last" call to TryClaim that returns true will automatically ensure subsequent claim calls return false.

"This is the last position. I'll return true, but set my state as stopped for future iterations".

File: sdks/go/pkg/beam/io/mongodbio/read.go [hunk 2]
```diff
@@ -442,53 +266,53 @@ func (fn *readFn) ProcessElement(
...
 
-		emit(value)
-	}
-
-	return cursor.Err()
-}
+		result := cursorResult{nextID: id}
+		if !rt.TryClaim(result) {
+			return cursor.Err()
+		}
```
Comment (lostluck) [inline sdks/go/pkg/beam/io/mongodbio/read.go:277]: I was going to comment that this should be *before* the work of decoding, but since we appear to not know which ID we're claiming until after decoding, this is fine.

Essentially, work should happen after a successful claim, but we're stuck with the API's we're given, and working non-ideal code is better than non-existent code that does what we need.

**Review by lostluck — APPROVED** [review]
As per usual, spectacularly well done. LGTM. I couldn't find anything to suggest.

You've been doing amazingly as a contributor for the Go SDK, with consistently high quality PRs, and being largely independant.  I would love to nominate you for a [Google Open Source Peer Bonus](https://opensource.google/documentation/reference/growing/peer-bonus) for your efforts.

If that's something you're interested in, could you email me at lostluck@apache.org with an email address I can provide to the nomination form, with "Peer Bonus" in the subject?

I have all the other information I would need to nominate you (name, github ID, why I'd nominate you).

Thanks again!

**Comment by johannaojeling** [comment]
> As per usual, spectacularly well done. LGTM. I couldn't find anything to suggest.
> 
> You've been doing amazingly as a contributor for the Go SDK, with consistently high quality PRs, and being largely independant. I would love to nominate you for a [Google Open Source Peer Bonus](https://opensource.google/documentation/reference/growing/peer-bonus) for your efforts.
> 
> If that's something you're interested in, could you email me at [lostluck@apache.org](mailto:lostluck@apache.org) with an email address I can provide to the nomination form, with "Peer Bonus" in the subject?
> 
> I have all the other information I would need to nominate you (name, github ID, why I'd nominate you).
> 
> Thanks again!

Wow thank you, that is so kind of you! I appreciate it and will send you an email. And thanks for the review and your comments!

Apologies for the PR being so big, hard to do one part without the other. This was an interesting task to work on, I might have a look at the open issue for allowing the SDF functions to have a context parameter and an error return type next.

**Comment by lostluck** [comment]
Email received!

No worries about PR length. Your code is well factored and very clear, so it wasn't hard to follow for me.
